src/Common/GWPAsan.cpp

```diff
@@ -0,0 +1,226 @@
+#include <Common/GWPAsan.h>
+
+#if USE_GWP_ASAN
+#    include <IO/ReadHelpers.h>
+#    include <gwp_asan/common.h>
+#    include <gwp_asan/crash_handler.h>
+#    include <gwp_asan/guarded_pool_allocator.h>
+#    include <gwp_asan/optional/options_parser.h>
+#    include <Common/ErrorCodes.h>
+#    include <Common/Exception.h>
+#    include <Common/Logger.h>
+#    include <Common/StackTrace.h>
+#    include <Common/logger_useful.h>
+
+#    include <atomic>
+#    include <iostream>
+
+namespace GWPAsan
+{
+
+namespace
+{
+size_t getBackTrace(uintptr_t * trace_buffer, size_t buffer_size)
+{
+    StackTrace stacktrace;
+    auto trace_size = std::min(buffer_size, stacktrace.getSize());
+    const auto & frame_pointers = stacktrace.getFramePointers();
+    memcpy(trace_buffer, frame_pointers.data(), trace_size * sizeof(uintptr_t));
+    return trace_size;
+}
+
+__attribute__((__format__ (__printf__, 1, 0)))
+void printString(const char * format, ...) // NOLINT(cert-dcl50-cpp)
+{
+    std::array<char, 1024> formatted;
+    va_list args;
+    va_start(args, format);
+
+    if (vsnprintf(formatted.data(), formatted.size(), format, args) > 0)
+        std::cerr << formatted.data() << std::endl;
+
+    va_end(args);
+}
+
+}
+
+gwp_asan::GuardedPoolAllocator GuardedAlloc;
+
+static bool guarded_alloc_initialized = []
+{
+    const char * env_options_raw = std::getenv("GWP_ASAN_OPTIONS"); // NOLINT(concurrency-mt-unsafe)
+    if (env_options_raw)
+        gwp_asan::options::initOptions(env_options_raw, printString);
+
+    auto & opts = gwp_asan::options::getOptions();
+    opts.Backtrace = getBackTrace;
+    opts.MaxSimultaneousAllocations = 256; /// for testing
```

Comment:
I will remove this

Reply:
should we enlarge the default `MaxSimultaneousAllocations`? for example 1024

Reply:
yes I wanted to discuss this, I wasn't sure about the defaults
we can easily fine tune them for our needs with env but it could be beneficial to have a working gwp asan for everyone

Reply:
I think (1024 * page_size * 2 < 10MB) is acceptable for memory size. 
We can run benchmark tests with different config to find a good balance.